S3 pre-signed URL: error with unicode filename #455

marxjohnson opened the pull request:
Rebased #456

marxjohnson commented:
@brendanheywood One of our Catalyst EU clients is waiting on this fix, would you mind taking a look now its been rebased?

brendanheywood commented on classes/local/store/s3/client.php:
```diff
@@ -486,9 +486,10 @@ public function generate_presigned_url($contenthash, $headers = array()) {
      * @return signed_url
      */
-    private function generate_presigned_url_s3($contenthash, $headers) {
+    private function generate_presigned_url_s3($contenthash, array $headers = [], $nicefilename = true) {
```
Can you add a unit test for this? It's unclear to see what was wrong with the urls it was creating and how they have changed

marxjohnson commented:
@brendanheywood I have added unit tests to the branch, and an additional test file for the "Pre-Signed URL Testing" test page. I rebased the branch to put the tests before Phuc's commit, so it's easy to run the tests with and without the change, and see the error when you try and download the file with the unicode character.

scara commented on tests/object_file_system_test.php:
```diff
@@ -585,6 +585,39 @@ public function test_can_generate_signed_url_by_hash_if_object_is_external() {
+        $filehash = $file->get_contenthash();
+        try {
+            $headers = [
+                    'Content-Disposition' => 'attachment; filename="😀.txt"',
```
Hello Everyone,
just lurking here 👀.
Is there any reason not to adopt even the [RFC 5987](https://www.rfc-editor.org/rfc/rfc5987#section-3.2) and [RFC 6266](https://www.rfc-editor.org/rfc/rfc6266#section-4.1) indications i.e.:
```
Content-Disposition: attachment; filename*=utf-8''%e2%82%ac%20rates
```
where the _filename_ there is something like `rawurlencode($originalfilename)`?
I mean something like:
```php
$result['filename'] = 'filename="' . utf8_encode($originalfilename) . '"'
                  . "; filename*=utf-8''" . rawurlencode($originalfilename);
```
Maybe it's not required by your use cases right now but it could be future proof.

HTH,
Matteo

marxjohnson replied:
Hi Matteo, you're right, that probably would make sense. With Moodle we can safely assume that the filename is UTF-8 encoded, so specifying that in the header would be correct. 

We are currently doing the rawurlencode() seperately, which doesn't quite follow the pattern you indicated. I will do an additional commit to rework it as you suggested.

scara replied:
> We should be OK just setting the filename as UTF-8 and doing `rawurlencode()`.

Modern browsers should be fine with that, though e.g. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition is not clear enough on the back-compat.
Old benchmarks like http://test.greenbytes.de/tech/tc2231/ give some hints but I think its fine to tell the browsers are now running fine in the modern days 😉.

HTH,
Matteo

tuanngocnguyen commented:
Reopen this PR as the following commit cause 403 error when retrieve files with pre-signed URL

[Set utf-8 encoding in signed URL filename header](https://github.com/catalyst/moodle-tool_objectfs/pull/493/commits/8e107e6f8eac3a92214d2c056493ae6d66ee91b1)
